SEGame UI Overhaul

Fabian-O01 opened the pull request.

Commit:
changed function name setRightGraphNodes to setGraphNodes and fixed its signature in gui.ts

Fabian-O01 commented:
@benkeks The following are my ideas on how to implement the six moves of the strong SEGame in the graph view. Before going ahead and writing code, it would be helpful to have a second opinion on some of those that we haven't talked about this Monday.

- observations: click on transition
- challenges: click on the processes that are supposed to belong to $Q_*$ and confirm selection with button (is there a better way to confirm than a button?)
- revivals: 
    - greyed out button that becomes available when move is possible
    - alternative: click on some $q_* \in Q_*$.
- answers: click on $q \in Q$
- pos. decisions:
    - simplest, but also "dirtiest" way: make _another_ button
    - alternative: provide something like two radio buttons with options "positive" and "negative" whenever a decision position arises. Clicking on an option executes the respective move.
    - probably the most intuitive solution: skip this move at first and execute with observation or challenge move
- neg. decisions: switch sides button (if not handled together with pos. decisions)

benkeks commented:
@Fabian-O01 I'm in meetings all day, today—so I can only give you a quick reply:

- Attacker moves
  - I agree that one would need a button for conjunction challenges. (Also, if $Q_* = \varnothing$!)
  - Positive/negative decisions: This should just be the attacker choosing between states p and q, shouldn't it? (So I don't think you need another button.)
- Defender moves:
  -  For revivals / conjunction answers, clicking on any q in the sets seems best. (At some point, it would be important to visualize that the $Q_*$ are selected collectively. But this is a secondary problem, I think.)

Generally, your ideas sound great!

Commit:
implemented input facilities for move selection from graph view in segame for all strong segame-moves except observations

Commit:
fixed several bugs in segame that would prevent the player from submitting moves, wrong coloring of processes in graph view and missing processes on game start

Commit:
added preview for move costs for segame

benkeks committed:
Hack a mechanism to detect edge label / action hover into handler & renderer

This leads to `hoverEdge` on `Handler` contain a JSON object with a reference to the hovered edge and the string of the hovered action.

benkeks commented:
@Fabian-O01 : I've pushed some code that makes the edges-labels (i.e. actions) currently hovered accessible from the `Handler`. (Hope this helps with creating a way of making them clickable and previewable.)

Commit:
made confirm challenge button in segame only show when relevant and fixed alignment when zooming in on canvas

Commit:
fixed edge label hitbox in handler and added onClick functionality to edge labels

Commit:
implemented observation move input functionality from se-game-graph and move cost preview. Removed move table entirely. Also another fix for hitboxes of edge labels in renderer.

Fabian-O01 committed:
Merge branch 'spectroscopy' into game-view-overhaul
